### DISCUSSION TITLE
Slow performance when converting from TIFF to HEIF

### DISCUSSION THREAD

**Discussion by fabriziogiudici** [ImageMagick/ImageMagick#6528] [Help]
### ImageMagick version

 7.1.1-3 Q16-HDRI x86_64 84d7ad135:20230311

### Operating system, version and so on

Docker image dpokidov/imagemagick:7.1.1.1-3-bullseye run under macOS 12.6.7

### Description

Hello.

I'm trying ImageMagick to convert TIFF files to HEIF. The process seems to be quite slow: e.g. a 24MP image takes on the average longer than one minute to complete.

The command line is:
   convert input.tiff -quality 90 output.heif

As you know, Docker under macOS runs in its own VM, which has been assigned 4 CPU and 12GB of RAM. The hardware is a
MacBook Pro (16-inch, 2019) — but I see that a single CPU is being used. I'd like to know whether the file format conversion is capable to exploit all available CPUs. If it doesn't I can work on my conversion script to process multiple files at a time (it's always a batch conversion).

TIFF images are ZIP compressed, exported from Capture One, coming from Sony a6600 files.
Thanks.

**Comment by snibgo** [comment]
How many images in your TIFF file?

> The process seems to be quite slow: e.g. a 24MP image takes on the average longer than one minute to complete.

My laptop (using SSD) takes under 3s for 35MP. There seems to be a problem on your system.

During your minute, what is happening? CPU saturation? Disk saturation?

IM doesn't have anything threadable when writing heif, but the serious work is delegated to libheif. I don't know if libheif is threadable.

What is the text output from "`magick -list resource`"? Perhaps something there is causing a problem.

**Reply by fabriziogiudici** [reply]
My TIFF files have a single main image and a lo-res preview, anyway I'm processing only the main image. Thanks for you answer because now, having a reference, I know there's something to investigate.
During processing I'm seeing CPU saturation. BTW, in the meantime I've modified my script to convert 12 images at the same time (my laptop has 16 CPUs) and I see that actually 12 CPUs are saturated. Here things might be affected by the fact that I don't know whether 16GB of RAM are enough for 12 images at the same time.

The output of the command is:

```
> magick -list resource
Resource limits:
  Width: 461.169PP
  Height: 461.169PP
  Area: 67.1496GP
  List length: unlimited
  Memory: 15.6345GiB
  Map: 31.269GiB
  Disk: unlimited
  File: 786432
  Thread: 12
  Throttle: 0
  Time: unlimited
```

Given that my current setup is complicated by Docker (that on macOS runs inside a VM), I'll prepare a benchmark on another laptop that runs Linux natively. I'll post here more info as soon as I get some meaningful figure.

**Comment by snibgo** [comment]
Nothing in "` magick -list resource`" would cause problems. Those are limits imposed by IM's `policy.xml`.

MacOS should be fine. I know nothing about VM or Docker. Maybe something there is causing problems.

**Comment by fabriziogiudici** [comment]
I did some test with better reproducibility (I found and fixed a thermal overload problem that caused underclocking). The typical 24MP file I have takes about 70s to be converted (a single i9 core engaged at 4.2GHz clock). Below I linked an example image.

The command used was:

`convert 20230719-0001.tif[0] -quality 90 heif:- > 20230719-0001@90%.heif`

If somebody would like to download the file (120MB) and measure the conversion time I'll be glad to have a reference figure to understand whether the performance of my system is fine.

Thanks.

[https://wetransfer.com/downloads/5f7272a34e8b438624bff69b75be6a5720230807182328/320c09391ff5998b9f0d8a224ee36af120230807182411/2404f8](https://wetransfer.com/downloads/5f7272a34e8b438624bff69b75be6a5720230807182328/320c09391ff5998b9f0d8a224ee36af120230807182411/2404f8?trk=TRN_TDL_01&utm_campaign=TRN_TDL_01&utm_medium=email&utm_source=sendgrid)
